### ISSUE TITLE
=> Bluesky: handle duplicate image blobs when some URLs aren't still serving

### ISSUE THREAD

**Issue by roddie-digital** [snarfed/bridgy-fed#1650]
Hello, I've recently noticed posts coming from a Mastodon account are sometimes being bridged but the included image is missing (it just shows a grey box in bsky). I think the first one I noticed was eleven days ago. They're not typified by being extra-large images (first one I checked was 128KB) and I can't see anything on its https://fed.brid.gy/ap/ page (the successful ones and the unsuccessful ones both show as "posted"). Going to the direct URL for the image just shows "Source image is unreachable". Any help in investigating or mitigating this issue would be much appreciated.

These are some example posts:

https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldxpydeza222
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldwaztjxpqf2
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldv7je7g5ky2
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldqsz76qh4k2
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldl5nowg3fu2
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldhz2rsfi5y2
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ldhewwd3vbv2
https://bsky.app/profile/ghostintheshell.roddie.social/post/3ld4oxi7q36f2

**Comment by snarfed** [comment]
Hmm, interesting bug, thanks for filing!

Images (blobs) in ATProto are content-addressed by their CID. In these cases, evidently you uploaded them to your Mastodon instance earlier, but didn't use them, and them uploaded them again later and used them in new posts. The first uploads had different URLs, and evidently eventually got garbage collected, but had the same CIDs, so our blobs still point to the earlier URLs.

Example: https://roddie.social/@ghostintheshell/113701671131291688 has image https://cdn.masto.host/roddiesocial/media_attachments/files/113/701/671/033/940/176/original/4c0e5692a6ced709.jpg . We stored our blob for that image at 2024-12-23 10:30:32 UTC with CID bafkreic4pr5u3pcea454drzc5gencumlqklwpt2uxo7n7oxyfaggl6z3ry.

However, we'd also seen that same image before, at URL https://cdn.masto.host/roddiesocial/media_attachments/files/113/527/096/578/357/345/original/f9fcd28bd82a59c6.jpg, and created a blob for it at that URL (with the same CID!) at 2024-11-22 15:12:59 UTC.

So, when a `getBlob` call comes in for that CID, we return the old URL, which isn't serving any more.

The fix will go somewhere around https://github.com/snarfed/arroba/blob/061d385ef4bd6dec18d818dce66a5ad65b113ab4/arroba/datastore_storage.py#L365-L371

**Comment by roddie-digital** [comment]
Thanks for figuring it out so quickly. Trying to get my head around what is happening here - the account is a bot that randomly selects images to post from a list of URLs in a CSV and posts without interaction are deleted after two weeks. The image would've always been used when it was uploaded because it's a single script that retrieves an image from a URL, uploads it to Mastodon to get a media ID and then creates a post referencing it. So it's when we get an image selected that has already been selected before but has been subsequently deleted? I'm surprised if it has the same ID because even if it randomly selects an image it's already selected, it still goes through the process of uploading it to Mastodon and getting a media ID.

**Comment by Tamschi** [comment]
ATProto seems to force blobs to be (account, content)-addressed _only_, i.e. the distinct ID gets entirely erased on the Bluesky side.

I think it's reasonable to serve the most recent URL seen for [that (DID, CID) pair](https://docs.bsky.app/docs/api/com-atproto-sync-get-blob).
It _does_ have to be keyed on the DID as well as the CID because otherwise it's trivial for anyone to post the same file and then take it offline, i.e. denial of service against other users.

**Comment by snarfed** [comment]
Sure! 

Bridgy Fed is already misbehaving here and not really complying with the protocol. Technically, we should be storing the blob ourselves and serving it directly, instead of redirecting to another URL. We've already caused minor problems for Bluesky themselves when external image URLs changed and served a different resolution or type.

I don't really want to take on the liability of hosting media, though, so I'm ok with this minor non-compliance and handling resulting issues like this one.

**Comment by Tamschi** [comment]
To be fair that is untrusted data even within their system, so they should be able to handle the file being switched (somewhat) gracefully.

**Comment by snarfed** [comment]
> To be fair that is untrusted data even within their system, so they should be able to handle the file being switched (somewhat) gracefully.

Definitely! And they do now. It's still a problem for us, though, since the image stops serving in Bluesky when that happens. 🤷

**Comment by snarfed** [comment]
We may also need a better approach than just using the latest URL for a given blob. In the example in https://github.com/snarfed/bridgy-fed/issues/1650#issuecomment-2561300784, the first URL stopped serving while the second URL is still ok, but that may not always be the case. We could easily see the opposite, ie the first URL serves ok, but the second one gets garbage collected.

We could keep all URLs we've seen for all blobs, and regularly re-check them and switch them all to URLs that still serve, but that's pretty heavyweight. Not sure of a better answer yet though.

**Comment by Tamschi** [comment]
> > To be fair that is untrusted data even within their system, so they should be able to handle the file being switched (somewhat) gracefully.
> 
> Definitely! And they do now. It's still a problem for us, though, since the image stops serving in Bluesky when that happens. 🤷

So they actually check the blob CID instead of just using it for addressing? Hm…

I suppose it _does_ help with non-custodial (i.e. the user signs each update) PDS hosting, yeah that's fair.

**Comment by snarfed** [comment]
Yup. https://github.com/snarfed/arroba/issues/35

**Comment by TomCasavant** [comment]
I assume this is the same issue so I'll just drop it in here for reference. If you scroll through the media on my bridged profile (https://bsky.app/profile/did:plc:glvkfxk3igeeybszoue3apxi) nearly every single one is a gray box. Though, with how often it's happening, I wouldn't be surprised if it's a configuration issue in my server.

**Comment by snarfed** [comment]
@TomCasavant ugh yeah, looks like a recent Bluesky team appview change broke us. Fix is merged, hopefully it'll be deployed soon! https://github.com/bluesky-social/atproto/pull/3370

**Comment by devinivy** [comment]
That fix should be deployed, please give me a tag if you're still seeing issues with the blob redirect!

**Comment by roddie-digital** [comment]
@devinivy Just seen a recurrence of it here:

https://bsky.app/profile/ghostintheshell.roddie.social/post/3lfvxj4usgg32

Is it just that any fedi posts deleted since the fix was deployed will display properly in bsky but older deletions can still cause a problem? Thanks

**Comment by snarfed** [comment]
@roddie-digital Bluesky's fix wasn't for this bug itself, it was for a different image bug on their end where they weren't handling `getBlob` redirects. The original bug here that your account triggers is still on us to fix in Bridgy Fed itself.

**Comment by roddie-digital** [comment]
Ah cool, thanks for the clarification

**Comment by snarfed** [comment]
^ that should hopefully fix this. @roddie-digital feel free to reopen if you see it recur!

**Comment by roddie-digital** [comment]
You legend, @snarfed ! I will keep an eye on it but even scrolling back through the posts on the acct bridged to bsky I can see images that were missing are now displayed. Thanks so much for this, I love Bridgy Fed

**Comment by snarfed** [comment]
Reopening, we're seeing this for @anujahooja / https://mastodon.social/users/quillmatiq now. We have two versions of his profile picture, blob `bafkreiggxmzguiblzbvsfpf44nhtvchlptauxf4uh2uaej5agc7zag3g24`:

* Pixelfed's, created and last updated 2025-01-15 18:06:27, now not serving: https://pxscdn.com/cache/avatars/685692430295721232/avatar_ajcy.jpg
* Mastodon's, 2024-06-08 02:29:48, older, still serving: https://files.mastodon.social/accounts/avatars/111/585/904/638/150/314/original/7d85eeef5bc753fc.jpeg

Ugh. The only way I see to fix this is to periodically re-check all blobs and see if they're still available. Not excited about that.

**Comment by quillmatiq** [comment]
I hate that, too. A half-step option: doing it when explicitly synced on the Bridgy Fed user page? We'll still run into this, but at least there will be an easy way for savvy users to handle it themselves. I also don't see this being a common issue, especially the case that we ran into with my accts.

**Comment by snarfed** [comment]
Another idea is to recheck them on demand as they're fetched, either on a constant frequency, eg if they were last checked over a day or a week or so ago, or based on their HTTP caching directives.

**Comment by snarfed** [comment]
Done! Did that last idea: we now re-fetch blobs every 3d to check that they're still serving, and if their contents have changed, we update our stored CID, MIME type, size, etc.
